Fix forward metrics for backwards compatibility

patdowney opened the pull request:
<!--
Thank you for contributing to CoreDNS!
Please provide the following information to help us make the most of your pull request:
-->

### 1. Why is this pull request needed and what does it do?
This pull request is necessary to fix monitoring for users of the commonly used `forward` plugin, that was broken by a PR I previously worked on to pull the `Proxy` functionality out of `forward` into somewhere more accessible.

This pull request adds the ability for consumers of `plugin/pkg/proxy`, like `forward` to specify an isolated metrics collector for the plugin.

### 2. Which issues (if any) are related?
#6160 

### 3. Which documentation changes (if any) need to be made?
Fixing this for backwards compatibility will reconcile the documentation that currently refers to `coredns_forward_` metrics.

### 4. Does this introduce a backward incompatible change or deprecation?
Hopefully not, it's fixing something I broker earlier!

patdowney committed:
fix: backward compatible forward plugin metrics

This change will both keeping backwards compatibility for `forward` metric
names, but also allowing any other consumers of the `proxy.Proxy` struct
to set bespoke metric subsystem names, if so desired.

Signed-off-by: Patrick Downey <patrick.downey@dioadconsulting.com>

SuperQ reviewed (changes requested):
I think the big issue here is that, in my opinion, metrics should not escape the plugin.

Rather than allow new metrics to be created with new names, we should add a third label to the metrics and require `NewProxy` require a name.

So something like:
```go
func NewProxy(addr, trans, proxyName string)
```

The the metrics would have names like:
```go
	requestCount = promauto.NewCounterVec(prometheus.CounterOpts{
		Namespace: plugin.Namespace,
		Subsystem: "proxy",
		Name:      "requests_total",
		Help:      "Counter of requests made per upstream.",
	}, []string{"proxy_name", "to"})
```

This way the metrics in one plugin don't leak into another.

patdowney committed:
This change introduces the `proxy_name` label to metrics exposed by `proxy.Proxy` plugin helper.

This is a breaking change and will remove the following metrics:
* `coredns_forward_healthcheck_failures_total{to, rcode}`
* `coredns_forward_conn_cache_hits_total{to, proto}`
* `coredns_forward_conn_cache_misses_total{to, proto}`

They are replaced with the following:
* `coredns_proxy_healthcheck_failures_total{proxy_name="forward", to, rcode}`
* `coredns_proxy_conn_cache_hits_total{proxy_name="forward", to, proto}`
* `coredns_proxy_conn_cache_misses_total{proxy_name="forward", to, proto}`

The following metrics have been kept, but should be considered deprecated so
they can be removed in the future:
`coredns_forward_requests_total{to}`
`coredns_forward_responses_total{to, rcode}`
`coredns_forward_request_duration_seconds{to, rcode}`

Signed-off-by: Patrick Downey <patrick.downey@dioadconsulting.com>

SuperQ commented on plugin/forward/metrics.go:
```diff
@@ -1,14 +1,38 @@
 
 // Variables declared for monitoring.
 var (
+	RequestCount = promauto.NewCounterVec(prometheus.CounterOpts{
```
Please make all of these metric vars unexported so they can't be used outside of the package.

```suggestion
	requestCount = promauto.NewCounterVec(prometheus.CounterOpts{
```

patdowney replied:
Have unexported all metrics defined within the `forward` plugin.

SuperQ commented on plugin/pkg/proxy/connect.go:
```diff
@@ -152,9 +152,9 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
-	RequestDuration.WithLabelValues(p.addr, rc).Observe(time.Since(start).Seconds())
+	RequestCount.WithLabelValues(p.proxyName, p.addr).Add(1)
+	RcodeCount.WithLabelValues(p.proxyName, p.addr, rc).Add(1)
+	RequestDuration.WithLabelValues(p.proxyName, p.addr, rc).Observe(time.Since(start).Seconds())
```
Same with the proxy metrics. While we're here let's fix them to be unexported.

```suggestion
	requestCount.WithLabelValues(p.proxyName, p.addr).Add(1)
	rcodeCount.WithLabelValues(p.proxyName, p.addr, rc).Add(1)
	requestDuration.WithLabelValues(p.proxyName, p.addr, rc).Observe(time.Since(start).Seconds())
```

patdowney replied:
Have unexported all metrics defined within the `proxy` plugin.

SuperQ commented on plugin/pkg/proxy/connect.go:
```diff
@@ -152,9 +152,9 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
-	RcodeCount.WithLabelValues(rc, p.addr).Add(1)
-	RequestDuration.WithLabelValues(p.addr, rc).Observe(time.Since(start).Seconds())
+	RequestCount.WithLabelValues(p.proxyName, p.addr).Add(1)
+	RcodeCount.WithLabelValues(p.proxyName, p.addr, rc).Add(1)
```
The `RequestCount` and `RcodeCount` metrics seem to be duplicates of each other, as well as duplicating the `_count` value of the `RequestDuration` metric.

Let's drop them as they are just extra cardinality data.

patdowney replied:
Have removed these and updated the `README.md` with pointers to derive the value from the `_count` value of the `requestDuration` metric.

SuperQ commented on plugin/forward/metrics.go:
```diff
@@ -22,3 +46,17 @@ var (
+		if !ok {
+			rc = strconv.Itoa(res.Rcode)
+		}
+		RcodeCount.WithLabelValues(addr, rc).Add(1)
```
Similar here, we don't actually need `RcodeCount` since we get the same values from the `RequestDuration` metric `_count` data.

We can remove this and mark it as a breaking change.

patdowney replied:
I've removed the duplicate metrics from `forward` completely.

SuperQ reviewed (changes requested):
Thanks, that is a good improvement.

Some other comments inline.

patdowney committed:
Remove redundant metrics from `forward` and `proxy` plugins.

The values of the retired:
* `coredns_proxy_requests_total`,
* `coredns_proxy_responses_total`,
* `coredns_forward_requests_total`,
* `coredns_forward_responses_total`

can be derived from `coredns_proxy_request_duration_seconds`

BREAKING CHANGE: Removes duplicate metrics from `forward` and `proxy` plugins

Signed-off-by: Patrick Downey <patrick.downey@dioadconsulting.com>

patdowney committed:
Don't export metrics from `forward` and `proxy`.

They're not needed outside of their modules.

Signed-off-by: Patrick Downey <patrick.downey@dioadconsulting.com>

patdowney committed:
Remove redundant `coredns_forward_request_duration_seconds` metric from `forward` plugin

The metric has been replaced by `coredns_proxy_request_duration_seconds{proxy_name="forward"}`

BREAKING CHANGE: Removes `coredns_forward_request_duration_seconds` metric from `forward` plugin, it is replaced by `coredns_proxy_request_duration_seconds{proxy_name="forward"}`

Signed-off-by: Patrick Downey <patrick.downey@dioadconsulting.com>

SuperQ reviewed (approved):
Looks good, thanks for the deprecated metric translation documentation.

Tantalor93 reviewed (approved):
LGTM, but we should make sure this is mentioned in release notes for 1.11.0 (https://github.com/coredns/coredns/pull/6135)
